### PR TITLE
Hw streams

### PR DESCRIPTION
**Pull request by Likh-Alex**

### CODE REVIEW

File: src/main/java/practice/StreamPractice.java [hunk 1]
```diff
@@ -23,7 +52,11 @@ public int findMinEvenNumber(List<String> numbers) {
      * Then return the average of all odd numbers or throw NoSuchElementException.
      */
     public Double getOddNumsAverage(List<Integer> numbers) {
-        return 0D;
+        return IntStream.range(0, numbers.size())
+                .map(number -> number % 2 == 0 ? numbers.get(number) : numbers.get(number) - 1)
+                .filter(number -> number % 2 != 0)
+                .average()
+                .orElseThrow(NoSuchElementException::new);
```
Comment (Ketch92) [inline src/main/java/practice/StreamPractice.java:59]: you can you only get() it throws needed exception by default

File: src/main/java/practice/StreamPractice.java [hunk 2]
```diff
@@ -50,7 +86,14 @@ public Double getOddNumsAverage(List<Integer> numbers) {
...
     public List<People> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<People> peopleList) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= fromAge
+                        && (person.getSex().equals(People.Sex.MAN)
+                        && person.getAge() <= maleToAge)
+                        || person.getAge() >= fromAge
+                        && (person.getSex().equals(People.Sex.WOMEN)
+                        && person.getAge() <= femaleToAge))
```
Comment (Ketch92) [inline src/main/java/practice/StreamPractice.java:95]: you can use the ternary operator here

File: src/main/java/practice/StreamPractice.java [hunk 3]
```diff
@@ -59,23 +102,12 @@ public Double getOddNumsAverage(List<Integer> numbers) {
...
-     * let's write our own impl of Predicate parametrized with Candidate in CandidateValidator.
-     */
-    public static List<String> validateCandidates(List<Candidate> candidates) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getSex().equals(People.Sex.WOMEN)
+                        && person.getAge() >= femaleAge)
+                .map(people -> people.getCats())
+                .flatMap(cats -> cats.stream())
+                .map(cat -> cat.getName())
```
Comment (Ketch92) [inline src/main/java/practice/StreamPractice.java:110]: you can use only one map operator

File: src/main/java/practice/StreamPractice.java [hunk 4]
```diff
@@ -1,11 +1,35 @@
...
+     * and live in urk for 10 years. For the last requirement use field periodsInUkr,
+     * which has following view:
+     * "2002-2015" For now we don't care if that was last 10 or not. We want to reuse our
+     * validation in future, so
+     * let's write our own impl of Predicate parametrized with Candidate in CandidateValidator.
+     */
+    public static List<String> validateCandidates(List<Candidate> candidates) {
+        return candidates.stream()
+                .filter(new CandidateValidator())
+                .map(candidate -> candidate.getName())
```
Comment (Ketch92) [inline src/main/java/practice/StreamPractice.java:28]: can you change the method reference?

File: src/main/java/practice/StreamPractice.java [hunk 5]
```diff
@@ -14,7 +38,12 @@
      * "Can't get min value from list: method_input_list"
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .flatMap(stringNum -> Arrays.stream(stringNum.split(",")))
+                .mapToInt(s -> Integer.parseInt(s))
```
Comment (Ketch92) [inline src/main/java/practice/StreamPractice.java:43]: also method reference can be applied

**Review by Ketch92 — CHANGES_REQUESTED** [review]

File: src/main/java/practice/StreamPractice.java (hunk 1 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: Done, thank you for an idea)

File: src/main/java/practice/StreamPractice.java (hunk 3 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: Done, refactored to use one map and one flatMap

File: src/main/java/practice/StreamPractice.java (hunk 2 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: I think in this particular case ternary operator shortens the code but complicates the readability. Anyway - done, it is good to practice)

File: src/main/java/practice/StreamPractice.java (hunk 4 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: Done, I just liked to use lambdas as they are clearer to me)

File: src/main/java/practice/StreamPractice.java (hunk 5 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: Done, as well)

**Review by Ketch92 — APPROVED** [review]

File: src/main/java/practice/StreamPractice.java [hunk 6]
```diff
@@ -14,7 +38,12 @@
      * "Can't get min value from list: method_input_list"
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .flatMap(stringNum -> Arrays.stream(stringNum.split(",")))
+                .mapToInt(Integer::parseInt)
+                .filter(num -> num % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list"));
```
Comment (RomanDubovskyi) [inline src/main/java/practice/StreamPractice.java:46]:
```suggestion
                .orElseThrow(() -> new RuntimeException("Can't get min value from list: " + numbers));
```
let's make our message more informative

**Review by RomanDubovskyi — APPROVED** [review]
fix message, rest seems to be good👍

File: src/main/java/practice/StreamPractice.java (hunk 6 above)
Comment (Likh-Alex) [reply src/main/java/practice/StreamPractice.java]: Done

**Pull request closed without merging**